HW3 - Jun Hwang
========================================================

You will have to analyze the RNA-seq data presented in:
Henn, A. D. et al. High-resolution temporal response patterns to influenza vaccine reveal a distinct human plasma cell gene signature. Scientific Reports 3, 2327 (2013).
  
1. Get the data from GEO. Please look at the class lecture slides as we've done it already


```{r, echo=FALSE}
setwd("C:/Users/Jun/Documents/GitHub/Biostat-578/")
```

```{r query-GEO, cache = TRUE}
library(GEOquery)
library(GEOmetadb)
# Download the mapping information and processed data
gd <- getGEO("GSE45735", destdir = "Data/GEO")
pd <- pData(gd[[1]])
# getGEOSuppFiles("GSE45735", makeDirectory=FALSE, baseDir = "Data/GEO/")
# The T14 file is problematic and needs to be fixed by hand
# Open the file, go to the bottom and remove the few inconsistent line at the end

# Note the regular expression to grep file names
files <- list.files(path = "Data/GEO/", pattern = "GSE45735_T.*.gz", full.names = TRUE)
file_list <- lapply(files, read.table, header=TRUE)
# Remove duplicated rows
file_list_unique <- lapply(file_list, function(x){x<-x[!duplicated(x$Gene),]; 
                                                  x <- x[order(x$Gene),]; 
                                                  rownames(x) <- x$Gene;
                                                  x[,-1]})
# Take the intersection of all genes
gene_list <- Reduce(intersect, lapply(file_list_unique, rownames))
file_list_unique <- lapply(file_list_unique, "[", gene_list,)
matrix <- as.matrix(do.call(cbind, file_list_unique))

# Clean up the pData
pd_small <- pd[!grepl("T13_Day8",pd$title),]
pd_small$Day <- sapply(strsplit(gsub(" \\[PBMC\\]", "", pd_small$title),"_"),"[",2)
pd_small$subject <- sapply(strsplit(gsub(" \\[PBMC\\]", "", pd_small$title),"_"),"[",1)
colnames(matrix) <- rownames(pd_small)
```

2. Use voom and limma to find genes that are differentially expressed at each time point compared to baseline (day 0). Use an FDR cutoff of 0.01.
Display your results using pheatmap showing the log fold-change of the differentially expressed genes grouped by time point.  


```{r}
library(limma)
new_set <- ExpressionSet(assayData = matrix+1)
pData(new_set) <- pd_small
design <- model.matrix(~subject+Day, new_set)
new_set_voom <- voom(new_set,design = design)
lm <- lmFit(new_set_voom, design)
eb <- eBayes(lm)

dd1 <- topTable(eb, coef="DayDay1", p.value=.01, number=Inf)
dd2 <- topTable(eb, coef="DayDay2", p.value=.01, number=Inf)
dd3 <- topTable(eb, coef="DayDay3", p.value=.01, number=Inf)
dd4 <- topTable(eb, coef="DayDay4", p.value=.01, number=Inf)
dd5 <- topTable(eb, coef="DayDay5", p.value=.01, number=Inf)
dd6 <- topTable(eb, coef="DayDay6", p.value=.01, number=Inf)
dd7 <- topTable(eb, coef="DayDay7", p.value=.01, number=Inf)
dd8 <- topTable(eb, coef="DayDay8", p.value=.01, number=Inf)
dd9 <- topTable(eb, coef="DayDay9", p.value=.01, number=Inf)
dd10 <- topTable(eb, coef="DayDay10", p.value=.01, number=Inf)
```

We see that Days 1 and 3 have 58 and 1 differentially expressed genes, respectively.

```{r}
#grab the names of those genes
gene_names <- c(rownames(dd1),rownames(dd3))
lfcmatrix <- matrix(NA,nrow=59,ncol=10)
for (i in 1:10){
  lfcmatrix[,i] <- topTable(eb, coef=paste0("DayDay",i), number=Inf)[gene_names, "logFC"]
}
rownames(lfcmatrix) <- gene_names
colnames(lfcmatrix) <- paste0("Day ", 1:10)
library(pheatmap)
pheatmap(lfcmatrix, cluster_cols=FALSE,cluster_rows=FALSE)
```
  
  
  
3. Perform a GSEA analysis using camera and the MSigDB Reactome pathway gene signatures. Display your results using pheatmap, again group by timepoint. This is similar to what we've done in class.


```{r}
library(GSEABase)
c2_set <- getGmt("./GSEA-sets/c2.all.v4.0.symbols.gmt")
gene_ids <- geneIds(c2_set)
sets_indices <- symbols2indices(gene_ids, rownames(new_set))
cont_matrix <- makeContrasts("DayDay1", levels=design)
res <- vector("list",length = 10)
for(i in 1:10)
  {
  contrast <- paste0("DayDay",i)
  cont_matrix <- makeContrasts(contrast, levels=design)
  res[[i]] <- camera(new_set_voom, sets_indices, design=design, contrast=cont_matrix, sort=FALSE)
}
PValue <- sapply(res, function(x){ifelse(x$Direction=="Up", -10*log10(x$PValue), 10*log10(x$PValue))})
rownames(PValue) <- rownames(res[[1]])
PValue_max <- rowMax(abs(PValue))
PValue_small <- PValue[PValue_max>30, ]
anno <- data.frame(Time=paste0("Day",1:10))
rownames(anno) <- colnames(PValue_small)  <- paste0("Day",1:10)
pheatmap(PValue_small, cluster_cols=FALSE)

```




